src/FetchInterceptor.js

```diff
@@ -0,0 +1,329 @@
+    this.config = { ...this.config, ...config };
+
+    if (!this.isConfigValid(this.config)) {
+      throw new Error(ERROR_INVALID_CONFIG);
+    }
+
+    this.accessTokenProvider = new AccessTokenProvider(this.fetch, this.config);
+  }
+
+  /**
+   * Authorizes fetch interceptor with given renew token
+   * @param refreshToken
+   * @param accessToken
+   */
+  authorize(refreshToken, accessToken) {
+    this.accessTokenProvider.authorize(refreshToken, accessToken);
+  }
+
+  /**
+   * Returns current authorization for fetch fetchInterceptor
+   * @returns {{accessToken: string, refreshToken: string}}
+   */
+  getAuthorization() {
+    return this.accessTokenProvider.getAuthorization();
+  }
+
+  /**
+   * Clears authorization tokens. Call this to effectively log out user from fetch interceptor.
+   */
+  clear() {
+    this.accessTokenProvider.clear();
+  }
+
+  /**
+   * Main intercept method, you should chain this inside wrapped fetch call
+   * @param args Args initially provided to fetch method
+   * @returns {Promise} Promise which resolves the same way as fetch would
+   */
+  intercept(...args) {
+    return new Promise((resolve, reject) => this.resolveIntercept(resolve, reject, ...args));
+  }
+
+  isConfigValid() {
+    return this.config.shouldIntercept &&
+      this.config.authorizeRequest &&
+      this.config.createAccessTokenRequest &&
+      this.config.parseAccessToken;
+  }
+
+  resolveIntercept(resolve, reject, ...args) {
+    const request = new Request(...args);
+    const { accessToken } = this.accessTokenProvider.getAuthorization();
+    const requestUnit = this.createRequestUnit(request, resolve, reject);
+
+    // if access token is not resolved yet
+    if (!accessToken) {
+      return this.accessTokenProvider
+        .renew()
+        .then(() => this.fetchWithRetry(requestUnit))
+        .catch(reject);
+    }
+
+    // attempt normal fetch operation
+    return this.fetchWithRetry(requestUnit)
+      .catch(reject);
+  }
+
+  fetchWithRetry(requestUnit) {
+    // prepare initial request unit
+    return Promise.resolve(requestUnit)
+      // resolve should intercept flag, when false, step is skipped
+      .then(this.shouldIntercept)
+      // authorize request
+      .then(this.authorizeRequest)
+      // last minute check if fetch should be performed
+      // this is as close as it gets to canceling events since
+      // fetch spec does not support cancel at the moment
+      .then(this.shouldFetch)
+      // perform fetch
+      .then(this.fetchRequest)
+      // check if response invalidates current access token
+      .then(this.shouldInvalidateAccessToken)
+      // perform token invalidation if neccessary
+      .then(this.invalidateAccessToken)
+      // handle unauthorized response by requesting a new access token and
+      // repeating a request
+      .then(this.handleResponse)
+      .catch(this.handleUnauthorizedRequest);
+  }
+
+  createRequestUnit(request, fetchResolve, fetchReject) {
```

Comment:
It wasn't clear to me what a request unit should be from its name. Maybe we should consider renaming it to something like `RequestContext` or possibly `RequestTask`.